FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpDecrypt.kt

```diff
@@ -55,7 +61,73 @@ object PgpDecrypt {
     }
   }
 
-  private fun processDecryptionException(e: Exception): Exception {
+  fun decryptWithResult(
+    srcInputStream: InputStream,
+    pgpSecretKeyRingCollection: PGPSecretKeyRingCollection,
+    protector: SecretKeyRingProtector
+  ): DecryptionResult {
+    srcInputStream.use { srcStream ->
+      val destOutputStream = ByteArrayOutputStream()
+      destOutputStream.use { outStream ->
+        if (srcStream.markSupported()) {
+          srcStream.mark(0)
+        }
+
+        try {
+          val decryptionStream = PGPainless.decryptAndOrVerify()
+            .onInputStream(srcStream)
+            .withOptions(
+              ConsumerOptions()
+                .addDecryptionKeys(pgpSecretKeyRingCollection, protector)
+            )
+
+          decryptionStream.use { it.copyTo(outStream) }
+          return DecryptionResult.withDecrypted(
+            content = destOutputStream,
+            filename = decryptionStream.result.fileName
+          )
+        } catch (e: Exception) {
+          if (e is DecryptionException && e.decryptionErrorType == DecryptionErrorType.NEED_PASSPHRASE) {
+            if (srcStream.markSupported()) {
+              srcStream.reset()
+              val keyIds = MessageInspector.determineEncryptionInfoForMessage(srcInputStream).keyIds
+              val fingerprints = mutableListOf<String>()
+              for (id in keyIds) {
+                var key: PGPKeyRing? = null
+                try {
+                  key = pgpSecretKeyRingCollection.getSecretKeyRing(id)
+                  protector.getDecryptor(id)
+                } catch (e: DecryptionException) {
+                  key?.let {
+                    val fingerprint = OpenPgpV4Fingerprint(key)
+                    fingerprints.add(fingerprint.toString())
+                  }
+                }
+              }
```

Comment:
This could be made a separate method - actually, PGPainless should have something like this maybe? In another milestone

Reply:
Agree :). You can read some details here https://github.com/pgpainless/pgpainless/issues/183. It's better than we have before but still not ideal.